### PR TITLE
Allow copying of browser-ready URLs for Internet streams.

### PR DESCRIPTION
**Pull request by coconutlulz**
This is the first step in allowing users to directly copy browser-ready URLs which link to a selected stream via the context menu.

I have introduced a new QAction and QUrl in InternetService, so that all services can use this in future.

The new feature has been integrated into the Soundcloud, SomaFM, Icecast and Digitally Imported services.

While I have tested the Digitally Imported functionality previously, this was before a premium account was required. As such, I cannot modify my changes to be in line with what I have for the other services, as I would not be able to test such changes. Would someone with a premium account be so kind as to test this? A commit to tidy up the DI code (i.e. the if else) would be most welcome.

This is my first ever pull request for free software, so go easy! :-)

Issue #4774

### CODE REVIEW

**Comment by hatstand** [comment]
Is there a plan for what to do for streams where we can't provide a browser-ready URL for some reason (e.g. Amazon)?

**Comment by coconutlulz** [comment]
If we can't provide one for the service, I'm not sure what can be done other than the default (i.e. don't provide the option).

**Comment by hatstand** [comment]
That's fine. As long as we don't provide URLs that don't work.

File: src/internet/digitally/digitallyimportedservicebase.cpp [hunk 1]
```diff
@@ -181,11 +183,59 @@ void DigitallyImportedServiceBase::ShowContextMenu(const QPoint& global_pos) {
...
+
+void DigitallyImportedServiceBase::ShowContextMenu(const QPoint& global_pos) {
+  EnsureMenuCreated();
+  QStandardItem* item = model()->itemFromIndex(model()->current_index());
+
+  if (item) {
+    int type = item->data(InternetModel::Role_Type).toInt();
+
+    // Digitally Imported streams have a Role_Type of 0.
+    if (type == 0) {
```
Comment (coconutlulz) [inline src/internet/digitally/digitallyimportedservicebase.cpp]: This conditional could be tidied up by replacing with `bool can_play`.

This bit of code was successfully tested **before** a premium account became mandatory, so any changes will need to be tested by someone with such an account.

**Comment by hatstand** [comment]
Run `make format`

**Comment by coconutlulz** [comment]
@hatstand Doesn't show me any problems; might be doing it incorrectly. Did I introduce incorrect formatting? Thanks.

**Comment by hatstand** [comment]
It's still wrong in `DigitallyImportedServiceBase::ShowContextMenu` at least

**Pull request closed without merging**